src/metabase/models/collection.clj

```diff
@@ -311,161 +480,173 @@
-                               (or (not (premium-features/enable-audit-app?))
-                                   (= read-or-write :write)))
-        has-permission? (fn [[id _]]
-                          (and
-                           (not (block-for-audit? id))
-                           (or (contains? permissions-set "/")
-                               (contains? ids-with-perm id))))]
-    (->> collection-id->collection
-         (filter has-permission?)
-         (merge root-map))))
-
-(def ^:private CollectionVisibilityConfig
-  [:map
-   [:include-trash-collection? {:optional true} :boolean]
-   [:include-archived-items {:optional true} [:enum :only :exclude :all]]
-   [:archive-operation-id {:optional true} [:maybe :string]]
-   [:permission-level {:optional true} [:enum :read :write]]])
-
-(defn- should-remove-for-archived? [include-archived-items collection]
-  (case include-archived-items
-    :all false
-    :exclude (:archived collection)
-    :only (not (or (:archived collection)
-                   (is-trash? collection)))))
-
-(defn- should-remove-for-trash? [include-trash-collection? collection]
-  (if include-trash-collection?
-    false
-    (is-trash? collection)))
-
-(defn- should-remove-for-archive-operation-id [archive-operation-id collection]
-  (and (some? archive-operation-id)
-       (not= archive-operation-id (:archive_operation_id collection))))
-
-(mu/defn permissions-set->visible-collection-ids :- VisibleCollections
-  "There are four knobs we need to take into account when turning the permissions set into visible collection IDs.
-  - permission-level: generally collections with `read` permission are visible. Sometimes we want to change this and only
-    view collections with `write` permissions.
-  - include-archived-items: are archived collections currently visible, or not?
-  - archive-operation-id: when looking at a archived collection, we want to restrict visible collections to those that share
-  that operation ID.
-  - include-trash-collection?: is the Trash Collection itself visible, or not? Sometimes we want to show the Trash but
-    not the archived items inside it."
-  ([permissions-set :- [:set :string]]
-   (permissions-set->visible-collection-ids permissions-set {}))
-  ([permissions-set :- [:set :string]
+(defn- should-display-root-collection?
+  "Should this user be shown the root collection, given the `visibility-config` passed?"
+  [visibility-config]
+  (and
+   ;; we have permission for it.
+   (can-access-root-collection? (:permission-level visibility-config))
+
+   ;; we're not *only* looking for archived items
+   (not= :only (:include-archived-items visibility-config))
+
+   ;; we're not looking for a particular `archive_operation_id`
+   (not (:archive-operation-id visibility-config))
+
+   ;; we're not looking for the children of a collection (root definitely isn't a child!)
+   (not (:effective-child-of visibility-config))))
+
+(mu/defn honeysql-filter-clause
```

Comment:
nit - I think the naming could be a little more specific. Maybe `collection-filter-clause` or `collection-permissions-filter-clause`? It's not a huge deal but `honeysql-filter-clause` could be referring to anything